lib/luaxtable.c

```diff
@@ -32,48 +14,150 @@
 	lunatik_object_t *runtime;
+	lunatik_object_t *skb;
 	union {
 		struct xt_match match;
 		struct xt_target target;
 	};
 	luaxtable_type_t type;
 } luaxtable_t;
 
-static bool luaxtable_match(const struct sk_buff *skb, struct xt_action_param *par)
+static struct {
+	lunatik_object_t *target;
+	lunatik_object_t *match;
+	bool match_fallback;
+	unsigned int target_fallback;
+} luaxtable_hooks = {NULL, NULL, false, XT_CONTINUE};
+
+static int luaxtable_invoke(lua_State *L, luaxtable_t *xtable, const char *op, int nargs, int nret)
 {
-	return true;
+	int ret = -1;
+	int base = lua_gettop(L) - nargs;
+
+	if (lunatik_getregistry(L, xtable) != LUA_TTABLE) {
+		pr_err("%s: could not find ops table\n", op);
+		goto err;
+	}
+
+	if (lua_getfield(L, -1, op) != LUA_TFUNCTION) {
+		pr_err("%s isn't defined\n", op);
+		goto err;
+	}
+
+	lua_insert(L, base + 1); /* op */
+	lua_pop(L, 1); /* table */
+
+	if (lua_pcall(L, nargs, nret, 0) != LUA_OK) {
+		pr_err("%s error: %s\n", op, lua_tostring(L, -1));
+		ret = -ECANCELED;
+		goto err;
+	}
+	if (nret != 0)
+		ret = luaL_optinteger(L, -1, 0);
+err:
+	return ret;
 }
 
-static int luaxtable_match_check(const struct xt_mtchk_param *par)
+static lunatik_object_t *luaxtable_getskb(lua_State *L, luaxtable_t *xtable)
 {
-	return 0;
+	if (lunatik_getregistry(L, xtable->skb) != LUA_TUSERDATA)
+		return NULL;
+
+	return (lunatik_object_t *)lunatik_toobject(L, -1);
 }
 
-static void luaxtable_match_destroy(const struct xt_mtdtor_param *par)
+static int luaxtable_domatch(lua_State *L, luaxtable_t *xtable, const struct sk_buff *skb, struct xt_action_param *par)
 {
+	lunatik_object_t *data = luaxtable_getskb(L, xtable);
+	if (data == NULL)
+		return -1;
+
+	if (skb_is_nonlinear(skb)) {
+		pr_err("skb is non_linear!\n");
+		return -1;
+	}
+
+	luadata_reset(data, skb->data, skb_headlen(skb), LUADATA_OPT_READONLY);
+	return luaxtable_invoke(L, xtable, "match", 1, 1);
 }
 
-static unsigned int luaxtable_target(struct sk_buff *skb, const struct xt_action_param *par)
+static int luaxtable_dotarget(lua_State *L, luaxtable_t *xtable, struct sk_buff *skb, const struct xt_action_param *par)
 {
-	return NF_ACCEPT;
+	lunatik_object_t *data = luaxtable_getskb(L, xtable);
+	if (data == NULL) {
+		pr_err("could not find skb\n");
+		return -1;
+	}
+
+	if (skb_is_nonlinear(skb)) {
+		pr_err("skb is non_linear!\n");
+		return -1;
+	}
+
+	luadata_reset(data, skb->data, skb_headlen(skb), LUADATA_OPT_KEEP);
+	return luaxtable_invoke(L, xtable, "target", 1, 1);
 }
 
-static int luaxtable_target_check(const struct xt_tgchk_param *par)
-{
-	return 0;
+#define LUAXTABLE_HOOK_CB(T, U, V, hook, huk) 				\
+static T luaxtable_##hook(U skb, V par)					\
+{									\
+	const luaxtable_info_t *info = (const luaxtable_info_t *)par->huk##info;	\
+	luaxtable_t *xtable = info->data;				\
+	int ret;							\
+									\
+	lunatik_run(xtable->runtime, luaxtable_do##hook, ret, xtable, skb, par);	\
+	return ret < 0 ? luaxtable_hooks.hook##_fallback : ret;		\
```

Comment:
I think we can pass `fallback` as an argument to `lunatik_run`, then avoiding this `ret < 0 ?` and unifying `LUAXTABLE_HOOK_CB` and `LUAXTABLE_DESTROYER_CB` in a single macro..

Reply:
I think it will be better to keep logically different sections to be seperate (destroy and hook), what do you suggest ?

Reply:
but why repeating code? you can have a `LUAXTABLE_CB` and define `LUAXTABLE_HOOK_CB` and `LUAXTABLE_DESTROYER_CB` using it..

```suggestion
	lunatik_run(xtable->runtime, cb, ret, xtable, skb, par, fallback);	\
	return ret;		\
```

Reply:
In the latest code, the number of parameters have changed, only 2 lines are similar, should I move them to a new macro ?